### ISSUE TITLE
KOPS wants to change encrypted volume on etcd on running cluster (but does nothing but complain everytime)

### ISSUE THREAD

**Issue by KlavsKlavsen** [kubernetes/kops#11062]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
kops 1.18.3

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
1.18.15

**3. What cloud provider are you using?**
aws

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
set cluster with etcd encrypted volume NOT defined.

When I do kops update cluster (after changing minsize to 3) it says:
```
Will modify resources:
  AutoscalingGroup/nodes.k8s.ops.example.com
        MinSize                  2 -> 3

  EBSVolume/a.etcd-events.k8s.ops.example.com
        Encrypted                true -> false

  EBSVolume/a.etcd-main.k8s.ops.example.com
        Encrypted                true -> false
```
If I run kops update cluster - without having changed anything - it shows the above 2 changes - but it never changes them - and I tried setting encryptedVolume: true - in config (as it appearently IS encrypted) - but thats now allowed.

**Comment by hakman** [comment]
Hi @KlavsKlavsen, could you check in the AWS console if those volumes are indeed encrypted or not?

**Comment by KlavsKlavsen** [comment]
All volumes under the instance is marked as Encrypted in AWs console.

**Comment by hakman** [comment]
In this case, I think you have enabled encryption by default or something similar:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html#encryption-by-default

At the moment kOps doesn't know this and it assumes that you had unencrypted volumes that later transformed into encrypted ones. The only way to fix this is to update the config to match reality.

**Comment by KlavsKlavsen** [comment]
@hakman I tried adding that using kops edit cluster - but that was rejected by kops (as not being changeable later on)

**Comment by rifelpet** [comment]
Maybe kops should change where it performs this immutable validation, rather than at the API layer it could ensure the encryption setting isnt changing closer to where it would otherwise call ec2.ModifyVolume.

This would mean the error would only happen if the setting would actually be changed on a volume, during `kops update cluster --yes` rather than during `kops edit cluster` or `kops replace`

**Comment by h3poteto** [comment]
I can reproduce it in master (git-e465d57d81).
Maybe we can not update clusters which are created by kops without `encryptedVolume` parameter when EBS default encryption is enabled. 

I will take this issue.

/assign

**Comment by behrtam** [comment]
Many orgs using AWS have a policy in place that will enforce default EBS encryption.

As @KlavsKlavsen mentioned if you try to edit the cluster and set the encryptedVolume it gets rejected: `[spec.etcdClusters[main].etcdMembers[a].encryptedVolume: Forbidden: encryptedVolume cannot be changed, spec.etcdClusters[events].etcdMembers[a].encryptedVolume: Forbidden: encryptedVolume cannot be changed]`

Starting with kops-1.20 that blocks us from updating. 
```$ kops-1.20 update cluster --yes
Using cluster from kubectl context: dev.example.tech

I0413 15:25:55.336111   63541 executor.go:111] Tasks: 0 done / 80 total; 43 can run
W0413 15:25:56.351113   63541 executor.go:139] error running task "EBSVolume/a.etcd-main.dev.example.tech" (9m58s remaining to succeed): Field cannot be changed: Encrypted
W0413 15:25:56.351159   63541 executor.go:139] error running task "EBSVolume/a.etcd-events.dev.example.tech" (9m58s remaining to succeed): Field cannot be changed: Encrypted
...
```

**Comment by h3poteto** [comment]
At the moment, AWS ignores `Encrypted` parameter in `ec2.CreateVolumeInput` when default encryption is enabled. So cluster config does not match, even if we specify `encryptedVolume: false` in cluster spec. Of course, kops creates volumes with `encryptedVolume: false` when we don't specify `encryptedVolume` in cluster spec, but the result does not change.

So I removed the validation, maybe you can edit your cluster using `kops edit` to change `encryptedVolume`.

**Comment by gswallow** [comment]
Can somebody explain @h3poteto's comment in plain English?  How does one get kops to create a cluster in an AWS account where EBS volume encryption has been enabled by default?

**Comment by KlavsKlavsen** [comment]
> Update. I think it means "edit the cluster spec file and add":

No. I didn't do that - I just created a cluster with KOPS - but as I did not specify disks to be encrypted - AWS sets them to be encrypted (perhaps we have a policy enforcing this in AWS somewhere)..  the validation removal should help.. I'll try once latest 1.20 version of kops is released with this.

**Comment by hakman** [comment]
@KlavsKlavsen I think 1.20.1 got released with the validation removal.

**Comment by JEFFA0101** [comment]
Just tried creating a cluster with version 1.20.1 and still receiving the same validation error.

**Comment by h3poteto** [comment]
@JEFFA0101 Could you tell me the procedure?

I can create a cluster with v1.20.1, and I can change `encryptedVolume` without validation error.

**Comment by AnubhavSabharwa** [comment]
add kmskeyId: arnofKMSkey in etcdClusters, this will resolved the issue
etcdMembers:
    - instanceGroup: instancGroupName
      encryptedVolume: true
      kmsKeyId: arn:aws:kms::489009450921:key/1903264b-0085-4ab0-9171-07a5379da793
      name: anyName

**Comment by AnubhavSabharwa** [comment]
> add kmskeyId: arnofKMSkey in etcdClusters, this will resolved the issue
> etcdMembers:
> - instanceGroup: instancGroupName
> encryptedVolume: true
> kmsKeyId: arn:aws:kms::accountId:key/kmskey
> name: anyName

This issue comes when you have encryption enabled on account level

**Comment by AnubhavSabharwa** [comment]
> @JEFFA0101 Could you tell me the procedure?
> 
> I can create a cluster with v1.20.1, and I can change `encryptedVolume` without validation error.

This issue comes when you have encryption enabled on account level

**Comment by JethroMV** [comment]
> kops 1.22
We just tried on 1.22 and it worked!
We had the above `encryptedVolume` set to `true`.
We also had a bad config:
```
 etcdClusters:
  - encryptedVolume: true
```
should have been:
```
etcdClusters:
  etcdMembers:
    encryptedVolume: true
```

So looks like this is working great with 1.22 and the correct encryption config.
